File: backend/src/pricing/pricing.service.ts
import { Injectable } from '@nestjs/common';
import { PrismaService } from '../prisma.service';
import { VehicleType } from '@prisma/client';

/// Pricing Service - Calculates fares with surge, night, area, waiting pricing
/// All pricing logic runs server-side for security and consistency
@Injectable()
export class PricingService {
  constructor(private prisma: PrismaService) {}

  /// Calculate fare for a trip
  async calculateFare(params: {
    pickupLat: number;
    pickupLng: number;
    dropoffLat: number;
    dropoffLng: number;
    vehicleType: VehicleType;
    distanceKm: number;
    durationMinutes: number;
    waitingMinutes?: number;
    promoCode?: string;
  }) {
    // 1. Get pricing config for vehicle type and zone
    const config = await this.getPricingConfig(params.vehicleType, params.pickupLat, params.pickupLng);

    // 2. Calculate base components
    const baseFare = config.baseFare;
    const distanceFare = params.distanceKm * config.perKmRate;
    const timeFare = params.durationMinutes * config.perMinuteRate;
    const subtotal = baseFare + distanceFare + timeFare;

    // 3. Check surge pricing
    const surgeMultiplier = await this.getSurgeMultiplier(params.pickupLat, params.pickupLng);
    const surgeCharge = subtotal * (surgeMultiplier - 1.0);

    // 4. Check night pricing
    const nightMultiplier = this.getNightMultiplier(config);
    const nightCharge = subtotal * (nightMultiplier - 1.0);

    // 5. Check area pricing (geofence)
    const areaMultiplier = await this.getAreaMultiplier(params.pickupLat, params.pickupLng);
    const areaCharge = subtotal * (areaMultiplier - 1.0);

    // 6. Calculate waiting charge
    const freeWaiting = config.freeWaitingMinutes;
    const chargeableWaiting = Math.max(0, (params.waitingMinutes || 0) - freeWaiting);
    const waitingCharge = chargeableWaiting * config.waitingFeePerMinute;

    // 7. Calculate promo discount
    let promoDiscount = 0;
    if (params.promoCode) {
      promoDiscount = await this.calculatePromoDiscount(params.promoCode, subtotal);
    }

    // 8. Calculate total
    const totalBeforeDiscount = subtotal + surgeCharge + nightCharge + areaCharge + waitingCharge;
    const totalFare = Math.max(config.minimumFare, totalBeforeDiscount - promoDiscount);

    return {
      baseFare,
      distanceFare,
      timeFare,
      surgeCharge: Math.round(surgeCharge * 100) / 100,
      nightCharge: Math.round(nightCharge * 100) / 100,
      areaCharge: Math.round(areaCharge * 100) / 100,
      waitingCharge: Math.round(waitingCharge * 100) / 100,
      cancellationFee: 0,
      promoDiscount: Math.round(promoDiscount * 100) / 100,
      totalFare: Math.round(totalFare * 100) / 100,
      surgeMultiplier,
      nightMultiplier,
      areaMultiplier,
      currency: config.currency,
      distanceKm: params.distanceKm,
      durationMinutes: params.durationMinutes,
      waitingMinutes: params.waitingMinutes || 0,
      vehicleType: params.vehicleType,
    };
  }

  /// Get pricing configuration for vehicle type and zone
  private async getPricingConfig(vehicleType: VehicleType, lat: number, lng: number) {
    // Check if location falls in a specific zone
    const zone = await this.prisma.$queryRaw`
      SELECT id FROM geo_fences
      WHERE type = 'SURGE_ZONE'
        AND "isActive" = true
        AND ST_Contains(
          polygon::geometry,
          ST_SetSRID(ST_MakePoint(${lng}, ${lat}), 4326)
        )
      LIMIT 1
    `;

    const zoneId = zone[0]?.id || null;

    const config = await this.prisma.pricingConfig.findFirst({
      where: { vehicleType, zoneId, isActive: true },
    });

    // Fallback to default config (no zone)
    if (!config) {
      const defaultConfig = await this.prisma.pricingConfig.findFirst({
        where: { vehicleType, zoneId: null, isActive: true },
      });
      return defaultConfig || this.getDefaultConfig(vehicleType);
    }

    return config;
  }

  /// Get surge multiplier based on demand/supply ratio in the area
  private async getSurgeMultiplier(lat: number, lng: number): Promise<number> {
    // Count active riders and available drivers in the area
    // This would use Redis for real-time counts
    // For now, return 1.0 (no surge)
    return 1.0;
  }

  /// Get night pricing multiplier
  private getNightMultiplier(config: any): number {
    const hour = new Date().getHours();
    const start = config.nightStartHour || 22;
    const end = config.nightEndHour || 6;

    if (start > end) {
      // Crosses midnight (e.g., 22:00 to 06:00)
      return (hour >= start || hour < end) ? config.nightMultiplier : 1.0;
    } else {
      return (hour >= start && hour < end) ? config.nightMultiplier : 1.0;
    }
  }

  /// Get area-based pricing multiplier from geofences
  private async getAreaMultiplier(lat: number, lng: number): Promise<number> {
    const fences = await this.prisma.$queryRaw`
      SELECT "pricingMultiplier" FROM geo_fences
      WHERE type IN ('SURGE_ZONE', 'AIRPORT', 'CUSTOM')
        AND "isActive" = true
        AND "pricingMultiplier" IS NOT NULL
        AND ST_Contains(
          polygon::geometry,
          ST_SetSRID(ST_MakePoint(${lng}, ${lat}), 4326)
        )
      ORDER BY "pricingMultiplier" DESC
      LIMIT 1
    `;

    return fences[0]?.pricingMultiplier || 1.0;
  }

  /// Calculate promo code discount
  private async calculatePromoDiscount(code: string, subtotal: number): Promise<number> {
    const promo = await this.prisma.promoCode.findUnique({ where: { code } });
    if (!promo || !promo.isActive) return 0;
    if (promo.validFrom > new Date() || promo.validTo < new Date()) return 0;
    if (promo.usageLimit && promo.usedCount >= promo.usageLimit) return 0;
    if (subtotal < promo.minFare) return 0;

    let discount = 0;
    if (promo.discountType === 'percentage') {
      discount = subtotal * (promo.discountValue / 100);
      if (promo.maxDiscount) discount = Math.min(discount, promo.maxDiscount);
    } else {
      discount = promo.discountValue;
    }

    return Math.min(discount, subtotal); // Can't discount more than subtotal
  }

  /// Default pricing configuration (fallback)
  private getDefaultConfig(vehicleType: VehicleType) {
    const multipliers: Record<string, number> = {
      MOTORCYCLE: 0.8,
      SEDAN: 1.0,
      SUV: 1.5,
      VAN: 1.8,
      LUXURY: 2.5,
    };

    const m = multipliers[vehicleType] || 1.0;

    return {
      baseFare: 5.0 * m,
      perKmRate: 1.5 * m,
      perMinuteRate: 0.25 * m,
      minimumFare: 10.0 * m,
      cancellationFee: 5.0,
      waitingFeePerMinute: 0.5,
      freeWaitingMinutes: 3,
      nightMultiplier: 1.3,
      nightStartHour: 22,
      nightEndHour: 6,
      currency: 'USD',
    };
  }
}

File: trippo_shared/lib/src/services/pricing_service.dart
import '../constants/app_constants.dart';
import '../models/pricing_model.dart';

/// Pricing Service - Client-side fare calculation and estimation
/// Works in conjunction with server-side pricing for verification
class PricingService {
  /// Calculate fare estimate based on distance, duration, and vehicle type
  /// This is a client-side estimation; the server calculates the final fare
  FareBreakdown calculateFareEstimate({
    required double distanceKm,
    required int durationMinutes,
    required String vehicleType,
    double baseFare = AppConstants.defaultBaseFare,
    double perKmRate = AppConstants.defaultPerKmRate,
    double perMinuteRate = AppConstants.defaultPerMinuteRate,
    double minimumFare = AppConstants.defaultMinimumFare,
    double surgeMultiplier = 1.0,
    double nightMultiplier = 1.0,
    double areaMultiplier = 1.0,
    int waitingMinutes = 0,
    double waitingFeePerMinute = AppConstants.defaultWaitingFeePerMinute,
    int freeWaitingMinutes = 3,
    double promoDiscount = 0.0,
    String currency = 'USD',
  }) {
    // 1. Get vehicle type multiplier
    final vehicleMultiplier = AppConstants.vehicleTypeMultipliers[vehicleType] ?? 1.0;

    // 2. Calculate base components
    final adjustedBaseFare = baseFare * vehicleMultiplier;
    final distanceFare = distanceKm * perKmRate * vehicleMultiplier;
    final timeFare = durationMinutes * perMinuteRate * vehicleMultiplier;

    // 3. Calculate subtotal before multipliers
    final subtotal = adjustedBaseFare + distanceFare + timeFare;

    // 4. Apply surge pricing
    final surgeCharge = subtotal * (surgeMultiplier - 1.0);

    // 5. Apply night pricing
    final nightCharge = subtotal * (nightMultiplier - 1.0);

    // 6. Apply area pricing
    final areaCharge = subtotal * (areaMultiplier - 1.0);

    // 7. Calculate waiting charges
    final chargeableWaitingMinutes = waitingMinutes > freeWaitingMinutes
        ? waitingMinutes - freeWaitingMinutes
        : 0;
    final waitingCharge = chargeableWaitingMinutes * waitingFeePerMinute;

    // 8. Calculate total before discount
    final totalBeforeDiscount = subtotal + surgeCharge + nightCharge + areaCharge + waitingCharge;

    // 9. Apply promo discount
    final finalTotal = (totalBeforeDiscount - promoDiscount).clamp(minimumFare, double.infinity);

    return FareBreakdown(
      baseFare: adjustedBaseFare,
      distanceFare: distanceFare,
      timeFare: timeFare,
      totalFare: finalTotal,
      surgeCharge: surgeCharge,
      nightCharge: nightCharge,
      areaCharge: areaCharge,
      waitingCharge: waitingCharge,
      cancellationFee: 0.0,
      promoDiscount: promoDiscount,
      surgeMultiplier: surgeMultiplier,
      nightMultiplier: nightMultiplier,
      areaMultiplier: areaMultiplier,
      currency: currency,
      distanceKm: distanceKm,
      durationMinutes: durationMinutes,
      waitingMinutes: waitingMinutes,
      vehicleType: vehicleType,
    );
  }

  /// Check if current time falls within night pricing hours
  bool isNightPricingActive({
    DateTime? currentTime,
    int nightStartHour = AppConstants.nightStartHour,
    int nightEndHour = AppConstants.nightEndHour,
  }) {
    final hour = (currentTime ?? DateTime.now()).hour;
    if (nightStartHour > nightEndHour) {
      // e.g., 22:00 to 06:00 - crosses midnight
      return hour >= nightStartHour || hour < nightEndHour;
    } else {
      return hour >= nightStartHour && hour < nightEndHour;
    }
  }

  /// Calculate surge multiplier based on demand/supply ratio
  double calculateSurgeMultiplier({
    required int activeRiders,
    required int availableDrivers,
  }) {
    if (availableDrivers == 0) return AppConstants.surgeMultiplierExtreme;
    
    final ratio = activeRiders / availableDrivers;
    
    if (ratio >= 5.0) {
      return AppConstants.surgeMultiplierExtreme;
    } else if (ratio >= 3.0) {
      return AppConstants.surgeMultiplierHigh;
    } else if (ratio >= 2.0) {
      return AppConstants.surgeMultiplierMedium;
    } else if (ratio >= 1.5) {
      return AppConstants.surgeMultiplierLow;
    }
    
    return 1.0; // No surge
  }

  /// Calculate cancellation fee based on trip state and timing
  double calculateCancellationFee({
    required String tripState,
    required DateTime tripCreatedAt,
    required double estimatedFare,
  }) {
    final minutesSinceCreation = DateTime.now().difference(tripCreatedAt).inMinutes;
    
    // Free cancellation within first 2 minutes
    if (minutesSinceCreation <= 2) return 0.0;
    
    // If driver already assigned and arriving
    if (tripState == 'DRIVER_ARRIVING' || tripState == 'DRIVER_ARRIVED') {
      return (estimatedFare * 0.3).clamp(
        AppConstants.defaultCancellationFee,
        estimatedFare * 0.5,
      );
    }
    
    // After driver assigned but before arriving
    if (tripState == 'DRIVER_ASSIGNED') {
      return AppConstants.defaultCancellationFee * 0.5;
    }
    
    return 0.0;
  }

  /// Format fare amount with currency symbol
  String formatFare(double amount, {String currency = 'USD'}) {
    final symbols = {
      'USD': '\$',
      'SAR': 'ر.س',
      'AED': 'د.إ',
      'PKR': '₨',
      'EUR': '€',
      'GBP': '£',
    };
    final symbol = symbols[currency] ?? currency;
    return '$symbol ${amount.toStringAsFixed(2)}';
  }
}
